api/src/main/java/org/apache/iceberg/metrics/ScanReport.java

```diff
@@ -102,40 +113,304 @@ public Builder withSnapshotId(long newSnapshotId) {
       return this;
     }
 
-    public Builder withFilter(Expression newFilter) {
-      this.filter = newFilter;
-      return this;
-    }
-
     public Builder withProjection(Schema newProjection) {
       this.projection = newProjection;
       return this;
     }
 
     public Builder fromScanMetrics(ScanMetrics newScanMetrics) {
-      this.scanMetrics = newScanMetrics;
+      this.scanMetrics = ScanMetricsResult.fromScanMetrics(newScanMetrics);
+      return this;
+    }
+
+    public Builder fromScanMetricsResult(ScanMetricsResult newScanMetricsResult) {
+      this.scanMetrics = newScanMetricsResult;
       return this;
     }
 
     public ScanReport build() {
       Preconditions.checkArgument(null != tableName, "Invalid table name: null");
-      Preconditions.checkArgument(null != filter, "Invalid expression filter: null");
       Preconditions.checkArgument(null != projection, "Invalid schema projection: null");
       Preconditions.checkArgument(null != scanMetrics, "Invalid scan metrics: null");
-      return new ScanReport(tableName, snapshotId, filter, projection, scanMetrics);
+      return new ScanReport(tableName, snapshotId, projection, scanMetrics);
+    }
+  }
+
+  /** A serializable version of a {@link Timer} that carries its result. */
+  public static class TimerResult implements Serializable {
+    private final String name;
+    private final TimeUnit timeUnit;
+    private final Duration totalDuration;
+    private final long count;
+
+    public TimerResult(String name, TimeUnit timeUnit, Duration totalDuration, long count) {
+      Preconditions.checkArgument(null != name, "Invalid timer name: null");
+      Preconditions.checkArgument(null != timeUnit, "Invalid time unit: null");
+      Preconditions.checkArgument(null != totalDuration, "Invalid duration: null");
+      this.name = name;
+      this.timeUnit = timeUnit;
+      this.totalDuration = totalDuration;
+      this.count = count;
+    }
+
+    public String name() {
+      return name;
+    }
+
+    public TimeUnit timeUnit() {
+      return timeUnit;
+    }
+
+    public Duration totalDuration() {
+      return totalDuration;
+    }
+
+    public long count() {
+      return count;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(name())
+          .add("duration", totalDuration())
+          .add("count", count)
+          .add("timeUnit", timeUnit)
+          .toString();
+    }
+
+    public static TimerResult fromTimer(Timer timer) {
```

Comment:
I find it odd that a static factory method is here. Is this auto-formatting, or can we move this above the constructor?

Reply:
moved it above the constructor